utils/benchmark.py

```diff
@@ -174,9 +181,20 @@ def print_performance_metrics(start_times: list, finish_times: list, num_runs: i
     else:
         assert len(start_times) == len(finish_times) == num_runs
 
+        variable_input_lengths_sum = num_runs - warm_up_runs
+        if variable_input_lengths is not None:
+            utils.print_warning_message("Performance results will be normalized due to variable input shape")
+            assert num_runs == len(variable_input_lengths)
+            average_input_length = statistics.mean(variable_input_lengths)
+            input_length_factors = [input_length / average_input_length for input_length in variable_input_lengths]
+            variable_input_lengths_sum = sum(variable_input_lengths[warm_up_runs:])
+
         latencies = []
         for i in range(warm_up_runs, num_runs):
-            latencies.append(finish_times[i] - start_times[i])
+            latency_sec = finish_times[i] - start_times[i]
+            if variable_input_lengths is not None:
+                latency_sec /= input_length_factors[i]
```

Comment:
Should we also divide it by the average_input_length? If I have it correctly, the unit here is seconds per run of average length, which when divided by average_input_length would become seconds per token. A bit later, `observed_throughput`'s unit is tokens per second, so there would be a symmetry here.

Reply:
No, the unit here is seconds as in the name of variable.
If there is some variation in input size across runs it should be feed to the function. Based on that factors are calculated and latency is normalized using them.

Example:
seq size = 384, latency = 3.2 sec
seq_size = 512, latency = 4.5 sec

Average of seq size is 448. Therefore:
seq size = 384, latency normalized = 3.2 / 0.8571 sec = 3.73 sec
seq suze 512, normalized = 4.5 / 1.1429 sec = 3.94 sec

Reply:
Throughput on the other hand has been changed to be calculated as observed instead of estimated from average latency.

Reply:
This example confirms what I was worried about. The normalized result depends on the average length of the sequence and will vary depending on the number of runs during the benchmark.

If we add a third sample:
seq size = 100, latency = 1.0 sec

we’ll get a different average sequence size and therefore different normalized latency:
Average of seq size is 332
seq size = 384, latency normalized = 3.2 / 1.1566 sec = 2.77  sec
seq size = 512, latency normalized = 4.5 / 1.5422 sec = 2.92 sec
seq size = 100, latency normalised = 1.0 / 0.3012 sec = 3.32 sec

But if we then divide the result by the average seq size, we get the same normalized result in both cases:

3.73 sec / 448 = 0.0083 sec
3.94 sec / 448 = 0.0088 sec

2.77  sec / 332 = 0.0083 sec
2.92 sec / 332 = 0.0088 sec
3.32 sec / 332 = 0.01 sec

We can also skip the step of calculating the ratios and just directly divide the latency by sequence size:

3.2 sec / 384 = 0.0083 sec
4.5 sec / 512 = 0.0088 sec
1.0 sec / 100 = 0.01 sec